### DISCUSSION TITLE
[code][refacto] asyncio

### DISCUSSION THREAD

**Discussion by artesim-zz** [Ortho4XP/Ortho4XP#29] [Ideas]
Ortho4XP already has some multiprocessing built in, for performance reasons.

It relies heavily on:
- multiple http requests to download external resources (OSM, elevation data and of course the tiles themselves)
- multiple external processes for encoding the tiles

In python 3, a lot of work has been done to implement Asynchronous I/O, which helps with working with asynchronous code while keeping the code predictable, and readable.

I've had very positive experience with the [trio library](https://github.com/python-trio/trio), and I think that the [refactoring effort](https://github.com/Ortho4XP/Ortho4XP/discussions/24) could include this idea.

**Comment by girotobial** [comment]
We absolutely should use async where possible. I've only done a surface level dive, but it's clear that Ortho4XP uses a lot of http calls to get the images from the relevant sources. 

I haven't used Trio before, so I'll have to give it a go. What advantages does it have over using plain asyncio with a library to handle async web requests like httpx?

**Reply by artesim-zz** [reply]
It's a little hard to explain, but trio enforces the dependency between async tasks through the structure of the code.
It all revolves around their notion of "nursery" (_questionable choice of name, and they did question it, but now it's set in stone..._).
Basically, when you do something like that:
```python
async def parent():
    async with trio.open_nursery() as nursery:
        nursery.start_soon(child1)
        nursery.start_soon(child_with_its_own_nested_nursery)
        nursery.start_soon(child2)
```

The 3 tasks will all be started immediately (the start_soon call doesn't block), but the "with" nursery block will not return until all the nested tasks are done. And a task can define and use its own nursery, which again will not exit until all its sub-tasks are done.

This is poorly explained, but that's roughly what I meant by "task dependency supported by the code structure".
More on that [in their design doc](https://trio.readthedocs.io/en/latest/design.html), its concise and explained way better than I did :)

Another pretty neat feature is how they handle exceptions raised by several tasks : they all bubble up the "tree of nurseries", and they're collected and dumped before the process exits.

In short, I find it a easier to work with trio, it feels more organized/less messy, and the design of their library seems sound.

Plus, there are bridges between base asyncio and trio, for libraries that are not written for it (and there's anyio, too, another library that bridges the two systems).

**Reply by artesim-zz** [reply]
Oh, and another advantage : I based the design of a small in-house application on trio, for work, which is now in production.
So I already went through the learning process with this library, and I know it was meeting all my needs.
